src/main/java/org/apache/commons/imaging/formats/jpeg/iptc/IptcParser.java

```diff
@@ -405,9 +417,25 @@ public PhotoshopApp13Data parsePhotoshopSegment(final byte[] bytes, final boolea
 
     public byte[] writeIPTCBlock(List<IptcRecord> elements)
             throws ImageWriteException, IOException {
+        Charset charset = DEFAULT_CHARSET;
+        for (final IptcRecord element : elements) {
+            final byte[] recordData = element.getValue().getBytes(charset);
+            if (!new String(recordData, charset).equals(element.getValue())) {
+                charset = StandardCharsets.UTF_8;
+                break;
+            }
+        }
         byte[] blockData;
         final ByteArrayOutputStream baos = new ByteArrayOutputStream();
         try (BinaryOutputStream bos = new BinaryOutputStream(baos, getByteOrder())) {
+            if( charset!=null && !charset.equals(DEFAULT_CHARSET) ) {
+                bos.write(IptcConstants.IPTC_RECORD_TAG_MARKER);
+                bos.write(IptcConstants.IPTC_ENVELOPE_RECORD_NUMBER);
+                bos.write(ENV_TAG_CODED_CHARACTER_SET);
+                byte[] codedCharset = charset.equals(StandardCharsets.UTF_8) ? CHARACTER_ESCAPE_SEQUENCE: charset.name().getBytes(StandardCharsets.ISO_8859_1);
+                bos.write2Bytes(codedCharset.length);
+                bos.write(codedCharset);
+            }
```

Comment:
Write encoded charset tag if charset is not null and ISO-8859-1. In the case of UTF-8, CHARACTER_ESCAPE_SEQUENCE is used, as done in general.